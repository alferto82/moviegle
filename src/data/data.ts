import { api_key, session_id } from "../utils/fetchUtils"

export const FAVORITE_URL =`https://api.themoviedb.org/3/account/acc/favorite/movies?api_key=${api_key}&session_id=${session_id}`
export const WATCH_LATER_URL =`https://api.themoviedb.org/3/account/acc/watchlist/movies?api_key=${api_key}&session_id=${session_id}`
export const POSTER_IMG_BASE= "https://image.tmdb.org/t/p/w200";

export interface IMovie {
    id: number,
    title: string, 
    release_date: string, 
    overview: string,
    poster_path: string
}

export interface IResponse<T> {
    results: Array<T>,
    page: number,
    total_pages: number
}

export type EventData = {
    movie_id: number
}