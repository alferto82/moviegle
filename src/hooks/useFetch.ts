import { useState, useEffect } from "react";
const useFetch = (url: string) => {
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const [currentPage, setCurrentPage] = useState(1);
  const [totalPages, setTotalPages] = useState(0);

  useEffect(() => {
    setLoading(true);
    fetch(`${url}&page=${currentPage < 1? 1: currentPage}`)
      .then((res) => res.json())
      .then((res) => {
        setData(res.results);
        setCurrentPage(res.page);
        setTotalPages(res.total_pages);
      })
      .catch(() => {
        setError(true);
      })
      .finally(() => setLoading(false));
  }, [currentPage]);
  
  return { data, loading, error, currentPage, totalPages, setCurrentPage };
};
export default useFetch;