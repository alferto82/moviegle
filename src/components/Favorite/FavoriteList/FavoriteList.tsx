import React,{ useEffect } from "react";
import { FAV_URL } from "../../../data/data";
import { favoriteEventChannel } from "../../../eventChannel/favorite";
import useFetch from "../../../hooks/useFetch";
import { Movies } from "../../movies/Movies";
import { Pagination } from "../../pagination/Pagination";
import { Spinner } from "../../Spinner/Spinner";


export const FavoriteList: React.FC = () => {
  const {data, loading, error, totalPages, currentPage, setCurrentPage} = useFetch(FAV_URL);

  useEffect(() => {
    // subscribe to events when mounting
    const unsubscribeOnRemoveFavoriteClick = favoriteEventChannel.on(
      'onRemoveFavoriteClick',
      (payload) => {
        console.log("removeFavoriteClick called", payload);
        setCurrentPage(0);
      }
    )
    // unsubscribe events when unmounting
    return () => {
      unsubscribeOnRemoveFavoriteClick()
    }
  }, [setCurrentPage])

  return (
    <div className="search_wrapper">
     {loading && <Spinner />}
     {data && !loading && ( <Movies favorite={false} watchLater={true} movies= {data}></Movies>)}
     {error && (<p>${error}</p>)}
     {totalPages > 1 && <Pagination page={currentPage} totalPages={totalPages} handlePagination={setCurrentPage}></Pagination> }
    </div>
  );
};
