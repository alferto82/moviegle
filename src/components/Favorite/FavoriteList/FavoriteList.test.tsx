import { queryByAttribute, render, screen } from '@testing-library/react';
import { FAV_URL, IMovie } from '../../../data/data';
import useFetch from '../../../hooks/useFetch';
import { FavoriteList } from './FavoriteList';


const mockMovie:IMovie = {
  id: 1,
  title: "mocked Movie", 
  release_date: "01/01/01", 
  overview: "mocked overview",
  poster_path: "mocked.jpg"
}

const mockResponse = {
  data: [], 
  loading: true, 
  error: false, 
  currentPage: 1,
  totalPages: 0, 
  setCurrentPage: jest.fn()
};

jest.mock('../../../hooks/useFetch');

const getByClass = queryByAttribute.bind(null, 'class');

test('Renders correctly', () => {
  (useFetch as jest.Mock).mockReturnValue(mockResponse);
  const component =render(<FavoriteList />);
  expect(useFetch).toHaveBeenCalledWith(FAV_URL);
  expect(screen.findByRole("Spinner")).not.toBeNull();
  
  expect(getByClass(component.container, "loaderWrapper")).not.toBeNull(); 
  expect(getByClass(component.container, "noMovies")).toBeNull(); 
  expect(getByClass(component.container, "pagination")).toBeNull(); 
});

test('Renders without loading', () => {
  (useFetch as jest.Mock).mockReturnValue({...mockResponse, loading: false});
  const component =render(<FavoriteList />);
  expect(useFetch).toHaveBeenCalledWith(FAV_URL);
  expect(screen.findByRole("Spinner")).not.toBeNull();

  expect(getByClass(component.container, "loaderWrapper")).toBeNull(); 
  expect(getByClass(component.container, "noMovies")).not.toBeNull(); 
  expect(getByClass(component.container, "pagination")).toBeNull(); 
});

test('Renders with results', () => {
  (useFetch as jest.Mock).mockReturnValue({...mockResponse, loading: false, data:[mockMovie], totalPages: 10});
  const component =render(<FavoriteList />);
  expect(useFetch).toHaveBeenCalledWith(FAV_URL);
  expect(screen.findByRole("Spinner")).not.toBeNull();

  expect(getByClass(component.container, "loaderWrapper")).toBeNull(); 
  expect(getByClass(component.container, "noMovies")).toBeNull(); 
  expect(getByClass(component.container, "pagination")).not.toBeNull(); 
});
