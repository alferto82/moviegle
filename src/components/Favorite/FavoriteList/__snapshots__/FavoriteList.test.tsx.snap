// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders loading correctly 1`] = `Promise {}`;
